discovery/watcher.go

```diff
@@ -239,31 +240,32 @@ func (w *Watcher) Stop() {
 	w.conn.Close()
 }
 
-func (w *Watcher) run() {
-	defer w.runComplete.SetDone()
-
+func (w *Watcher) run(bo backoff.BackOff, nextServer func(*addrSet) error) {
```

Comment:
**Optional, please feel free to take or leave** ☺️ 

I often find that needing to "stub out" a private member (in this case the `nextServer` method) for testing means something isn't quite right in the interface boundaries. In other words, if it's not possible to test something as a "black box" by interrogating its public API and observing its side effects with mocks etc, there might be too much logic in one place.

The "side effect" I'd be looking to observe here is that `<reconnection logic>` happens without delay, indicating the backoff has been reset correctly.

I think there'd be a couple of ways of observing this with an outside-in approach:
1. Running a "real" gRPC server that implements the server API, having it return an error, and checking that the client reconnects (i.e. the server receives another feature discovery RPC) without delay.
2. Moving the connection logic into a `Connector` type we can mock out.

I think the second approach would make the most sense. Both could be implemented without adding real sleeping to the tests by coordinating between the mock/server and test using channels.

Reply:
Ah, I completely agree, with the caveat that I think low-level unit tests still have their place. For example, we have the block box test that checks the Watcher will reconnect on error + the unit test I've added here that checks an aspect of that reconnection (backoff + reset), which gives us a level of confidence without needing more black box tests which I find, at least, can be sort of harder to understand and debug. If that makes sense 🙂

I totally agree there are test and interface improvement opportunities here though. When I get some time, I'll follow up on this in another PR, if that's okay?

Reply:
Cool, sounds good — definitely not a blocker for this PR 👍🏻